pubsub/fakepubsub/fake.go

```diff
@@ -0,0 +1,209 @@
+// This should not be used for production: it is intended for local
+// development.
+//
+// fakepubsub does not support any types for As.
+package fakepubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/google/go-cloud/pubsub/driver"
+)
+
+type topic struct {
+	mu        sync.Mutex
+	subs      []*subscription
+	nextAckID int
+	closed    bool
+}
+
+// CreateTopic creates a new topic.
+// Create subscribers for the topic before publishing.
+func CreateTopic() driver.Topic {
+	return &topic{}
+}
+
+// SendBatch sends the messages to all the subscriptions for this topic. It is an
+// error if the topic is closed or has no subscriptions.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// Check for closed or canceled before doing any work.
+	if t.closed {
+		return errors.New("fakepubsub: SendBatch: topic closed")
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	// Catch the common mistake of publishing before any subscriptions have been created.
+	if len(t.subs) == 0 {
+		return errors.New("fakepubsub: SendBatch: no subscriptions for topic")
+	}
+	// Associate ack IDs with messages here. It would be a bit better if each subscription's
+	// messages had their own ack IDs, so we could catch one subscription using ack IDs from another,
+	// but that would require copying all the messages.
+	for i, m := range ms {
+		m.AckID = t.nextAckID + i
+	}
+	t.nextAckID += len(ms)
+	for _, s := range t.subs {
+		s.add(ms)
+	}
+	return nil
+}
+
+// Close closes the topic. Subsequent calls to SendBatch will fail.
+func (t *topic) Close() error {
+	t.mu.Lock()
+	t.closed = true
+	t.mu.Unlock()
+	return nil
+}
+
+type subscription struct {
+	mu          sync.Mutex
+	topic       driver.Topic
+	ackDeadline time.Duration
+	msgs        map[driver.AckID]*message // all unacknowledged messages
+	ctx         context.Context           // for Close
+	cancel      func()
+}
+
+// CreateSubscription creates a new subscription for the given topic.
+// Unacknowledged messages will become available for redelivery after ackDeadline.
+// By default, real time (as determined by time.Now) is used. Set the variable Now
+// to a different function to mock time.
+func CreateSubscription(t driver.Topic, ackDeadline time.Duration) driver.Subscription {
+	tt := t.(*topic)
+	ctx, cancel := context.WithCancel(context.Background())
+	s := &subscription{
+		topic:       tt,
+		ackDeadline: ackDeadline,
+		msgs:        map[driver.AckID]*message{},
+		ctx:         ctx,
+		cancel:      cancel,
+	}
+	tt.mu.Lock()
+	tt.subs = append(tt.subs, s)
+	tt.mu.Unlock()
+	return s
+}
+
+type message struct {
+	msg        *driver.Message
+	expiration time.Time
+}
+
+func (s *subscription) add(ms []*driver.Message) {
+	s.mu.Lock()
+	for _, m := range ms {
+		// The new message will expire at the zero time, which means it will be
```

Comment:
Does an expiry time of zero mean that the message never expires?

Reply:
No, it means the message expired a long time ago. The zero time in Go is an actual time in the past.